In [81]:
# third party
import numpy
import pandas
import numpy as np
from tabulate import tabulate

In [64]:
data = pandas.read_csv('olympics.csv', index_col=0, skiprows=1)
description = data.describe()
print(tabulate(description, headers="keys", tablefmt="orgtbl"))

|       |   № Summer |      01 ! |     02 ! |      03 ! |    Total |   № Winter |   01 !.1 |   02 !.1 |   03 !.1 |   Total.1 |   № Games |    01 !.2 |   02 !.2 |    03 !.2 |   Combined total |
|-------+------------+-----------+----------+-----------+----------+------------+----------+----------+----------+-----------+-----------+-----------+----------+-----------+------------------|
| count |  147       |  147      |  147     |  147      |   147    |  147       | 147      | 147      | 147      |  147      |  147      |  147      |   147    |  147      |           147    |
| mean  |   13.4762  |   65.4286 |   64.966 |   69.7959 |   200.19 |    6.70068 |  13.0476 |  13.034  |  12.898  |   38.9796 |   20.1769 |   78.4762 |    78    |   82.6939 |           239.17 |
| std   |    7.07236 |  405.55   |  399.31  |  427.187  |  1231.31 |    7.43319 |  80.7992 |  80.6344 |  79.5884 |  240.917  |   13.257  |  485.013  |   478.86 |  505.855  |          1469.07 |
| min   |    1       |    0      | 

In [65]:
data.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

In [66]:
data.tail()

№ Summer  01 !  02 !  03 !  \
Independent Olympic Participants (IOP) [IOP]         1     0     1     2   
Zambia (ZAM) [ZAM]                                  12     0     1     1   
Zimbabwe (ZIM) [ZIM]                                12     3     4     1   
Mixed team (ZZX) [ZZX]                               3     8     5     4   
Totals                                              27  4809  4775  5130   

                                              Total  № Winter  01 !.1  02 !.1  \
Independent Olympic Participants (IOP) [IOP]      3         0       0       0   
Zambia (ZAM) [ZAM]                                2         0       0       0   
Zimbabwe (ZIM) [ZIM]                              8         1       0       0   
Mixed team (ZZX) [ZZX]                           17         0       0       0   
Totals                                        14714        22     959     958   

                                              03 !.1  Total.1  № Games  \
Independent Olympic Participants (IOP) [IOP]       0        0        1   
Zambia (ZAM) [ZAM]                                 0        0       12   
Zimbabwe (ZIM) [ZIM]                               0        0       13   
Mixed team (ZZX) [ZZX]                             0        0        3   
Totals                                           948     2865       49   

                                              01 !.2  02 !.2  03 !.2  \
Independent Olympic Participants (IOP) [IOP]       0       1       2   
Zambia (ZAM) [ZAM]                                 0       1       1   
Zimbabwe (ZIM) [ZIM]                               3       4       1   
Mixed team (ZZX) [ZZX]                             8       5       4   
Totals                                          5768    5733    6078   

                                              Combined total  
Independent Olympic Participants (IOP) [IOP]               3  
Zambia (ZAM) [ZAM]                                         2  
Zimbabwe (ZIM) [ZIM]                                       8  
Mixed team (ZZX) [ZZX]                                    17  
Totals                                                 17579

In [67]:
print((data["Combined total"].sum() - data.loc["Totals"]["Combined total"]) * 2)
print(data["Combined total"].sum())
print(data["Combined total"]

SyntaxError: unexpected EOF while parsing (<ipython-input-67-85598e5eb101>, line 3)

In [68]:
new_columns = ("summer_participations",
	       "summer_gold",
	       "summer_silver",
	       "summer_bronze",
	       "summer_total",
	       "winter_participations",
	       "winter_gold",
	       "winter_silver",
	       "winter_bronze",
	       "winter_total",
	       "total_participations",
	       "total_gold",
	       "total_silver",
	       "total_bronze",
	       "total_combined")
assert len(new_columns) == len(data.columns)

column_remap = dict(zip(data.columns, new_columns))

In [69]:
#Check Remaping
if data.columns[0] != "summer_participations":
    for column in data.columns:
        data.rename(columns={column:column_remap[column]}, inplace=True)

data.head(1)

summer_participations  summer_gold  summer_silver  \
Afghanistan (AFG)                     13            0              0   

                   summer_bronze  summer_total  winter_participations  \
Afghanistan (AFG)              2             2                      0   

                   winter_gold  winter_silver  winter_bronze  winter_total  \
Afghanistan (AFG)            0              0              0             0   

                   total_participations  total_gold  total_silver  \
Afghanistan (AFG)                    13           0             0   

                   total_bronze  total_combined  
Afghanistan (AFG)             2               2

In [70]:
# split the index by '('
names_ids = data.index.str.split('\s\(')
data.index = names_ids.str[0]
data['ID'] = names_ids.str[1].str[:3]
data = data.drop('Totals')

Index([          'AFG)',           'ALG)',           'ARG)',           'ARM)',
           'ANZ) [ANZ]', 'AUS) [AUS] [Z]',           'AUT)',           'AZE)',
                 'BAH)',           'BRN)',
       ...
                 'UZB)',           'VEN)',           'VIE)',           'ISV)',
           'YUG) [YUG]',     'IOP) [IOP]',     'ZAM) [ZAM]',     'ZIM) [ZIM]',
           'ZZX) [ZZX]',              nan],
      dtype='object', length=147)
Float64Index([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
              ...
              nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
             dtype='float64', length=147)
Float64Index([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
              ...
              nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
             dtype='float64', length=147)


In [71]:
data.head()

summer_participations  summer_gold  summer_silver  summer_bronze  \
Afghanistan                     13            0              0              2   
Algeria                         12            5              2              8   
Argentina                       23           18             24             28   
Armenia                          5            1              2              9   
Australasia                      2            3              4              5   

             summer_total  winter_participations  winter_gold  winter_silver  \
Afghanistan             2                      0            0              0   
Algeria                15                      3            0              0   
Argentina              70                     18            0              0   
Armenia                12                      6            0              0   
Australasia            12                      0            0              0   

             winter_bronze  winter_total  total_participations  total_gold  \
Afghanistan              0             0                    13           0   
Algeria                  0             0                    15           5   
Argentina                0             0                    41          18   
Armenia                  0             0                    11           1   
Australasia              0             0                     2           3   

             total_silver  total_bronze  total_combined   ID  
Afghanistan             0             2               2  AFG  
Algeria                 2             8              15  ALG  
Argentina              24            28              70  ARG  
Armenia                 2             9              12  ARM  
Australasia             4             5              12  ANZ

What is the first country in in the data-frame?

In [77]:
def answer_zero():
    return data.iloc[0].name
answer_zero()

'Afghanistan'

Which country has won the most gold medals in summer games? 

In [87]:
def answer_one():
    return data.summer_gold.argmax()
answer_one()

C:\Users\simra\AppData\Local\Programs\Python\Python36\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  


'United States'

Which country had the biggest difference between their summer and winter gold medal counts?

In [88]:
def answer_two():
    return (data.summer_gold - data.winter_gold).abs().argmax()

answer_two()

C:\Users\simra\AppData\Local\Programs\Python\Python36\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  


'United States'

Which country has the biggest difference between their summer gold medal counts and winter gold medal counts relative to their total gold medal count?


$$ \frac{Summer~Gold - Winter~Gold}{Total~Gold} $$


Only include countries that have won at least 1 gold in both summer and winter.

In [91]:
def answer_three():
    elegible = data[(data.summer_gold>=1) & (data.winter_gold>=1)]
    ratios = (elegible.summer_gold - elegible.winter_gold).abs()/elegible.total_gold
    return ratios.argmax()

answer_three()

C:\Users\simra\AppData\Local\Programs\Python\Python36\lib\site-packages\ipykernel_launcher.py:15: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  from ipykernel import kernelapp as app


'Bulgaria'

Write a function that creates a Series called "Points" which is a weighted value where each gold medal(Gold.2) counts for 3 points, silver medals (Silver.2) for 2 points, and bronze medals (Bronze.2) for 1 point. The function should return only the column (a Series object) which you created.

In [92]:
def answer_four():
    points = numpy.zeros(len(data))
    points += data.total_gold * 3
    points += data.total_silver * 2
    points += data.total_bronze
    return pandas.Series(points, index=data.index)

In [97]:
points = answer_four()
print(points)
assert points.loc["United States"] == 5684

Afghanistan                            2.0
Algeria                               27.0
Argentina                            130.0
Armenia                               16.0
Australasia                           22.0
Australia                            923.0
Austria                              569.0
Azerbaijan                            43.0
Bahamas                               24.0
Bahrain                                1.0
Barbados                               1.0
Belarus                              154.0
Belgium                              276.0
Bermuda                                1.0
Bohemia                                5.0
Botswana                               2.0
Brazil                               184.0
British West Indies                    2.0
Bulgaria                             411.0
Burundi                                3.0
Cameroon                              12.0
Canada                               846.0
Chile                                 24.0
China      

In [99]:
census_data = pandas.read_csv('census.csv')
census_data.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
0        4779736            4780127          4785161  ...          0.002295   
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
0         -0.193196          0.381066          0.582002         -0.467369   
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
0     1.030015     0.826644     1.383282     1.724718     0.712594  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  

[5 rows x 100 columns]

In [100]:
class CensusVariables:
    state_name = "STNAME"
    county_name = "CTYNAME"
    census_population = "CENSUS2010POP"
    region = "REGION"
    population_2014 = "POPESTIMATE2014"
    population_2015 = "POPESTIMATE2015"
    population_estimates = ["POPESTIMATE2010",
			    "POPESTIMATE2011",
			    "POPESTIMATE2012",
			    "POPESTIMATE2013",
			    population_2014,
			    population_2015]
    county_level = 50
    summary_level = "SUMLEV"

In [106]:
counties = census_data[census_data[
    CensusVariables.summary_level]==CensusVariables.county_level]
#print(counties)
# this throws off the numeric index for the argmax method so reset it
counties = counties.reset_index()
#print(counties)
# but the last question wants the original index
counties_original_index = census_data[census_data[
    CensusVariables.summary_level]==CensusVariables.county_level]
print(counties_original_index)

      SUMLEV  REGION  DIVISION  STATE  COUNTY     STNAME             CTYNAME  \
1         50       3         6      1       1    Alabama      Autauga County   
2         50       3         6      1       3    Alabama      Baldwin County   
3         50       3         6      1       5    Alabama      Barbour County   
4         50       3         6      1       7    Alabama         Bibb County   
5         50       3         6      1       9    Alabama       Blount County   
6         50       3         6      1      11    Alabama      Bullock County   
7         50       3         6      1      13    Alabama       Butler County   
8         50       3         6      1      15    Alabama      Calhoun County   
9         50       3         6      1      17    Alabama     Chambers County   
10        50       3         6      1      19    Alabama     Cherokee County   
11        50       3         6      1      21    Alabama      Chilton County   
12        50       3         6      1   

Which state has the most counties in it? (hint: consider the sumlevel key carefully! You'll need this for future questions too…)

In [111]:
def answer_five():

    return counties.groupby(
	CensusVariables.state_name).count().COUNTY.argmax()

In [112]:
answer_five()

C:\Users\simra\AppData\Local\Programs\Python\Python36\lib\site-packages\ipykernel_launcher.py:10: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  # Remove the CWD from sys.path while we load stuff.


'Texas'

Only looking at the three most populous counties for each state, what are the three most populous states (in order of highest population to lowest population)? Use CENSUS2010POP.

In [113]:
def answer_six():
    top_threes = counties.groupby(
	CensusVariables.state_name
    )[CensusVariables.census_population].nlargest(3)
    states = top_threes.groupby(level=0).sum()
    return list(states.nlargest(3).index)

answer_six()

['California', 'Texas', 'Illinois']

Which county has had the largest absolute change in population within the period 2010-2015? (Hint: population values are stored in columns POPESTIMATE2010 through POPESTIMATE2015, you need to consider all six columns.)

e.g. If County Population in the 5 year period is 100, 120, 80, 105, 100, 130, then its largest change in the period would be |130-80| = 50.

In [121]:
def answer_seven():
    seven = counties.iloc[(counties[CensusVariables.population_estimates].max(axis=1) - counties[CensusVariables.population_estimates].min(axis=1)).argmax()]
    return seven[CensusVariables.county_name]

answer_seven()

C:\Users\simra\AppData\Local\Programs\Python\Python36\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: 
The current behaviour of 'Series.argmax' is deprecated, use 'idxmax'
instead.
The behavior of 'argmax' will be corrected to return the positional
maximum in the future. For now, use 'series.values.argmax' or
'np.argmax(np.array(values))' to get the position of the maximum
row.
  


'Harris County'

In this datafile, the United States is broken up into four regions using the "REGION" column.

Create a query that finds the counties that belong to regions 1 or 2, whose name starts with 'Washington', and whose POPESTIMATE2015 was greater than their POPESTIMATE2014.

In [124]:
def answer_eight():
    regions = counties_original_index[
	(counties_original_index[CensusVariables.region]==1) |
	(counties_original_index[CensusVariables.region]==2)]
    washingtons = regions[
	regions[CensusVariables.county_name].str.startswith("Washington")]
    grew = washingtons[washingtons[CensusVariables.population_2015] >
		       washingtons[CensusVariables.population_2014]]
    return grew[[CensusVariables.state_name,
		 CensusVariables.county_name]]

outcome = answer_eight()
print(outcome)

            STNAME            CTYNAME
896           Iowa  Washington County
1419     Minnesota  Washington County
2345  Pennsylvania  Washington County
2355  Rhode Island  Washington County
3163     Wisconsin  Washington County
